includes/libraries/parsecsv.lib.php

```diff
@@ -0,0 +1,1163 @@
+     *
+     * @return Processed value
+     */
+    protected function _enclose_value($value = null, $delimiter = null) {
+        if (is_null($delimiter)) {
+            $delimiter = $this->delimiter;
+        }
+        if ($value !== null && $value != '') {
+            $delimiter_quoted = preg_quote($delimiter, '/');
+            $enclosure_quoted = preg_quote($this->enclosure, '/');
+            if (preg_match("/" . $delimiter_quoted . "|" . $enclosure_quoted . "|\n|\r/i", $value) || ($value{0} == ' ' || substr($value, -1) == ' ') || $this->enclose_all) {
+                $value = str_replace($this->enclosure, $this->enclosure . $this->enclosure, $value);
+                $value = $this->enclosure . $value . $this->enclosure;
+            }
+        }
+
+        return $value;
+    }
+
+    /**
+     * Check file data
+     *
+     * @access protected
+     * @param   file   local filename
+     *
+     * @return  true or false
+     */
+    protected function _check_data($file = null) {
+        if (empty($this->file_data)) {
+            if (is_null($file)) {
+                $file = $this->file;
+            }
+
+            return $this->load_data($file);
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if passed info might be delimiter
+     * Only used by find_delimiter
+     *
+     * @access protected
+     * @param  [type] $char      [description]
+     * @param  [type] $array     [description]
+     * @param  [type] $depth     [description]
+     * @param  [type] $preferred [description]
+     *
+     * @return special string used for delimiter selection, or false
+     */
+    protected function _check_count($char, $array, $depth, $preferred) {
+        if ($depth == count($array)) {
+            $first = null;
+            $equal = null;
+            $almost = false;
+            foreach ($array as $key => $value) {
+                if ($first == null) {
+                    $first = $value;
+                } elseif ($value == $first && $equal !== false) {
+                    $equal = true;
+                } elseif ($value == $first + 1 && $equal !== false) {
+                    $equal = true;
+                    $almost = true;
+                } else {
+                    $equal = false;
+                }
+            }
+
+            if ($equal) {
+                $match = ($almost) ? 2 : 1;
+                $pref = strpos($preferred, $char);
+                $pref = ($pref !== false) ? str_pad($pref, 3, '0', STR_PAD_LEFT) : '999';
+
+                return $pref . $match . '.' . (99999 - str_pad($first, 5, '0', STR_PAD_LEFT));
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Read local file
+     *
+     * @access protected
+     * @param   file   local filename
+     *
+     * @return  Data from file, or false on failure
+     */
+    protected function _rfile($file = null) {
+        if (is_readable($file)) {
```

Comment:
Needs a `strlen($input) <= PHP_MAXPATHLEN && ` to fix warnings when csv paths are longer than allowed by the PHP configuration